File: hyunjae/프로그래머스/고득점키트/다리를_지나는_트럭.java
import java.util.*;

class Solution {
    class Timer {
        int weight;
        int time;

        public Timer(int weight, int time) {
            this.weight = weight;
            this.time = time;
        }
    }
    
    public int solution(int bridge_length, int weight, int[] truck_weights) {
        int answer = 0;
        Queue<Timer> timer = new LinkedList<>();
        int nowWeight = 0;
        int index = 0;

        while (index < truck_weights.length || !timer.isEmpty()) {
            // 시간 경과
            answer++;

            // 다리에서 트럭 제거
            if (!timer.isEmpty() && timer.peek().time == answer) {
                nowWeight -= timer.poll().weight;
            }

            // 새로운 트럭 추가
            if (index < truck_weights.length && nowWeight + truck_weights[index] <= weight) {
                timer.add(new Timer(truck_weights[index], answer + bridge_length));
                nowWeight += truck_weights[index];
                index++;
            }
        }
        
        return answer;
    }
}

File: sejeong/다리를 지나는 트럭.py
def solution(bridge_length,weight,truck_weights):
    second = 0
    current_w = 0  # 현재 다리의 무게
    queue = [0] * bridge_length  # 다리

    while truck_weights:
        second += 1
        current_w -= queue.pop(0)
        if current_w + truck_weights[0] <= weight:  # 다리에 트럭이 들어갈 수 있으면
            queue.append(truck_weights[0]) # 다리에 트럭넣고
            current_w += truck_weights.pop(0) # 현재무가증가시키기
        else:
            queue.append(0) # 못들어가면 0넣기

    second += len(queue) # 마지막 트럭이 들어가면 나올때까지 시간더해주기
    return second


bridge_length = 2
weight = 10
truck_weights = [7,4,5,6]
print(solution(bridge_length,weight,truck_weights))